src/avisi/atlassian/connect/firestore.clj

```diff
@@ -0,0 +1,195 @@
+(ns avisi.atlassian.connect.firestore
+  (:require
+    [mount.core :as m :refer [defstate]]
+    [com.fulcrologic.guardrails.core :refer [>defn >def => ?]]
+    [clojure.spec.alpha :as s]
+    [camel-snake-kebab.core :as csk]
+    [avisi.atlassian.connect.jwt :as jwt]
+    [cambium.core :as log]
+    [clojure.string :as string])
+  (:import
+    (com.google.cloud.firestore
+      FirestoreOptions
+      FirestoreOptions$Builder
+      Firestore
+      DocumentReference
+      CollectionReference
+      DocumentSnapshot
+      Query
+      FieldPath
+      QueryDocumentSnapshot
+      QuerySnapshot)
+    (com.google.api.core ApiFuture)
+    (java.util HashMap ArrayList)
+    (java.time Duration)
+    (java.util Date UUID)
+    (com.google.cloud Timestamp)))
+
+(set! *warn-on-reflection* true)
+
+(>def ::collection-path string?)
+(>def ::document-path string?)
+(>def ::document map?)
+(>def ::host-client-key string?)
+(>def ::collection string?)
+(>def ::field string?)
+(>def ::value string?)
+
+(defn transform-map-keys [m f]
+  (reduce (fn [acc [k v]] (assoc acc (f k) (cond->> v (or (vector? v) (instance? ArrayList v)) (mapv f)))) {} m))
+
+(defn keys-as-string [m] (transform-map-keys m str))
+
+(defn string-keys-to-keywords [m]
+  (transform-map-keys
+    m
+    (fn [s]
+      (cond-> s
+        (string/starts-with? s ":")
+          (->
+            (subs 1)
+            (keyword))))))
+
+(defn timestamps-to-java-dates [m]
+  (reduce (fn [acc [k v]] (assoc acc k (cond->> v (instance? Timestamp v) (.toDate)))) {} m))
+
+(defstate options
+  :start
+  (->
+    (FirestoreOptions/newBuilder)
+    ^FirestoreOptions$Builder (.setProjectId (:firestore-project-id (m/args)))
+    (.build)))
+
+(defstate service :start ^Firestore (.getService ^FirestoreOptions options) :stop (.close ^Firestore service))
+
+(>defn store-document!
+  [{::keys [document-path document]}]
+  [(s/keys :req [::document-path ::document]) => any?]
+  (->
+    ^Firestore service
+    ^DocumentReference (.document document-path)
+    ^ApiFuture (.set (keys-as-string document))
+    (.get)))
+
+(defn doc-ref->doc [^DocumentReference doc-ref]
+  (->
+    doc-ref
+    ^ApiFuture (.get)
+    ^DocumentSnapshot (.get)
+    (.getData)))
+
+(defn doc-ref->entity [^DocumentReference doc-ref]
+  (some->>
+    (doc-ref->doc doc-ref)
+    (into {})
+    string-keys-to-keywords
+    timestamps-to-java-dates))
+
+(>defn fetch-document
+  [{::keys [document-path]}]
+  [(s/keys :req [::document-path]) => (? map?)]
+  (->
+    ^Firestore service
+    ^DocumentReference (.document document-path)
+    doc-ref->entity))
+
+(defn host-client-key [host] (:atlassian-connect.host/client-key host))
+
+(>defn store-host!
+  [{::keys [document]}]
+  [(s/keys :req [::document]) => any?]
+  (store-document!
+    {::document-path (format "hosts/%s" (host-client-key document))
```

Comment:
Ik zag in de andere firstore api's dat ze hier een vector van maken dus: `[hosts (host-client-key document)]` dat vond ik persoonlijk wat mooier dan strings